packages/api/app/launchpadSaml.js

```diff
@@ -0,0 +1,127 @@
+  const certificate = JSONPath('$..ds:X509Certificate', metadata);
+  if (certificate) return flatten(certificate);
+  return [];
+};
+
+/**
+ * Store the SAML response's token in the AccessResponse table and return a JWT
+ * from the derived values.
+ *
+ * @param {Object} samlResponse - post_assert response from saml IDP provider
+ *
+ * @returns {Promise<Object>} - a valid JWT token that can be used for authentication.
+ */
+const buildLaunchpadJwtToken = async (samlResponse) => {
+  const {
+    user: { name_id: username, session_index: accessToken }
+  } = samlResponse;
+
+  const expirationTime = Date.now() + 60 * 60 * 1000;
+  const accessTokenModel = new AccessToken();
+  await accessTokenModel.create({ accessToken, expirationTime, username });
+  return createJwtToken({ accessToken, expirationTime, username });
+};
+
+/**
+* convenience function to set up SAML Identity and Service Providers
+*/
+const prepareSamlProviders = async () => {
+  const LaunchpadX509Certificate = await launchpadPublicCertificate(
+    process.env.LAUNCHPAD_METADATA_PATH
+  );
+
+  const spOptions = {
+    entity_id: process.env.ENTITY_ID,
+    assert_endpoint: process.env.ASSERT_ENDPOINT,
+    force_authn: false,
+    nameid_format: 'urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified',
+    sign_get_request: false,
+    allow_unencrypted_assertion: true
+  };
+
+  const idpOptions = {
+    sso_login_url: process.env.IDP_LOGIN,
+    sso_logout_url: null,
+    certificates: LaunchpadX509Certificate
+  };
+
+  const idp = new saml2.IdentityProvider(idpOptions);
+  const sp = new saml2.ServiceProvider(spOptions);
+
+  return { idp, sp };
+};
+
+// Starting point for SAML SSO login
+const login = async (req, res) => {
+  // saml2-js stuff
+  const { idp, sp } = await prepareSamlProviders();
+  const relayState = req.query.RelayState;
+  sp.create_login_request_url(
+    idp,
+    { relay_state: relayState },
+    (err, loginUrl) => {
+      if (err != null) return res.boom.badRequest('Could not create login request url.', err);
+      return res.redirect(loginUrl);
+    }
+  );
+};
+
+// SAML AssertionConsumerService (ACS) endpoint.
+const auth = async (req, res) => {
+  const { idp, sp } = await prepareSamlProviders();
+  sp.post_assert(idp, { request_body: req.body }, (err, samlResponse) => {
+    if (err != null) {
+      return res.boom.badRequest('SAML post assert error', err);
+    }
+    return buildLaunchpadJwtToken(samlResponse)
+      .then((LaunchpadJwtToken) => {
+        const Location = `${req.body.RelayState}/?token=${LaunchpadJwtToken}`;
+        return res.redirect(Location);
+      })
+      .catch((error) => res.boom.badRequest('Could not build JWT from SAML response', error));
+  });
+};
+
+const notImplemented = async (req, res) => res.boom.notImplemented(
+  `endpoint: "${req.path}" not implemented. Login with launchpad.`
+);
+
+const tokenEndpoint = notImplemented;
+const refreshEndpoint = notImplemented;
```

Comment:
So this is disabled with Launchpad auth because there is no effective way to refresh tokens with SAML authentication?

I see that we are setting the token expiration time to one hour above, meaning that after an hour the dashboard will go to do a token refresh, fail, and then log the user out. That's not an unreasonable experience, just clarifying that is what would happen. The only other option is increasing the token expiration time

Reply:
So I don't know for sure if there's a way to refresh SAML tokens.  But right now, the token endpoint always sends you to build an earthdata or google oAuth2 provider. and we don't want to allow that if we're Launchpad protected.  I have a ticket to see if we can implement tokens for saml Launchpad and I will think about a refresh when I look at that.

https://bugs.earthdata.nasa.gov/browse/CUMULUS-1574

Reply:
Yeah, I should have been more clear. I wasn't so much asking a question as confirming my understanding. I don't think there is any way to do refresh with SAML, so this is fine